primitives/src/merkle_tree/internal.rs

```diff
@@ -895,3 +927,89 @@ where
         }
     }
 }
+
+/// Iterator type for a merkle tree
+pub struct MerkleTreeIter<'a, E: Element, I: Index, T: NodeValue> {
+    stack: Vec<&'a MerkleNode<E, I, T>>,
+}
+
+impl<'a, E: Element, I: Index, T: NodeValue> MerkleTreeIter<'a, E, I, T> {
+    /// Initialize an iterator
+    pub fn new(root: &'a MerkleNode<E, I, T>) -> Self {
+        Self { stack: vec![root] }
+    }
+}
+
+impl<'a, E, I, T> Iterator for MerkleTreeIter<'a, E, I, T>
+where
+    E: Element,
+    I: Index,
+    T: NodeValue,
+{
+    type Item = (&'a I, &'a E);
+
+    fn next(&mut self) -> Option<Self::Item> {
```

Comment:
Do I understand correct usage which would be analogous to previous `fn leaves()` (for serialization) would be to `match` the returned `Option` to filter empty fields? Or, since it is an `Iter` we would do `mt.iter().filter_map(|item|item)`. I'm specifically thinking of the case of the sparse merkle tree.

Reply:
It's an `Iter`. So just use `mt.iter().collect()` or `mt.into_iter().collect()` depending on whether you want to consume the merkle tree ownership.
All empty leaves are filtered out by default.